_posts/2020-08-24-native-loader-net5.md

```diff
@@ -0,0 +1,38 @@
+---
+layout: blog
+title: "Native Library Loading in .NET 5"
+author: Ryan Lucia
+tags: [loader]
+---
+
+After years of work, Mono can now be built out of the [dotnet/runtime repository](https://github.com/dotnet/runtime) in a .NET 5-compatible mode! This mode means numerous changes in the available APIs, managed and embedding, as well as internal runtime behavioral changes to better align Mono with CoreCLR and the .NET ecosystem.
+
+One area with multiple highly impactful changes to the runtime internals is library loading. For managed assemblies, Mono now follows the algorithms outlined on [this page](https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/loading-managed), which result from the removal of `AppDomain`s and the new `AssemblyLoadContext` APIs. The only exception to this is that Mono still supports bundles registered via the embedding API, and so the runtime will check that after the active `AssemblyLoadContext` cache.
+
+The managed loading changes are fairly clear and well documented, but unmanaged library loading has changed in numerous ways, some of them far more subtle.
+
+First off, Mono now follows [this algorithm](https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/loading-unmanaged) for P/Invoke resolution. Notably absent here is DllMap's XML config files, which have been disabled entirely in .NET 5. The DllMap embedding APIs are also disabled by default on desktop platforms, but can be enabled by building with `--enable-minimal=dllmap`.
+
+In place of DllMap, users are encouraged to utilize the [NativeLibrary](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=netcore-3.1) resolution APIs, which are set in managed code, and the [runtime hosting properties](https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting#step-3---prepare-runtime-properties), which are set by embedders with the `monovm_initialize` function. If these are insufficient for your use case, please [tell us about it](https://github.com/dotnet/runtime/issues/37213)! We're always looking to improve our interop functionality, and in particular with .NET 6 will be evaluating `NativeLibrary`, so please speak up.
+
+A more subtle, yet no less impactful change is that native library loading now defaults to `RTLD_LOCAL` to be consistent with CoreCLR and Windows, as opposed to our historical behavior of `RTLD_GLOBAL`. This change prevents symbol collision, and will both break and enable various scenarios/libraries. What this means in practice is that on Unix-like platforms, libraries are no longer loaded into a single global namespace and when looking up symbols, the library must be correctly specified.
+
+Historically, in Mono on Linux it was possible to load library `foo` containing symbol `bar`, and then invoke `bar` with a P/Invoke like so: 
+```
+[DllImport("asdf")]
```

Comment:
nit: add `// note the incorrect library name` on this line to highlight the problem